source/extensions/http/cache/file_system_http_cache/insert_context.cc

```diff
@@ -210,94 +108,138 @@ void InsertOperationQueue::insertHeaders(std::shared_ptr<InsertOperationQueue> p
       });
 }
 
-void InsertOperationQueue::insertBody(std::shared_ptr<InsertOperationQueue> p,
-                                      const Buffer::Instance& chunk,
-                                      InsertCallback ready_for_next_chunk, bool end_stream) {
-  push(p, QueuedFileChunk{QueuedFileChunk::Type::Body, chunk, ready_for_next_chunk, end_stream});
+void FileInsertContext::insertBody(const Buffer::Instance& fragment,
+                                   InsertCallback ready_for_next_fragment, bool end_stream) {
+  absl::MutexLock lock(&mu_);
+  if (!cleanup_) {
+    // Already cancelled, do nothing, return failure.
+    ready_for_next_fragment(false);
+    return;
+  }
+  ASSERT(!cancel_action_in_flight_, "should be no actions in flight when receiving new data");
+  callback_in_flight_ = ready_for_next_fragment;
+  size_t sz = fragment.length();
+  Buffer::OwnedImpl consumable_fragment(fragment);
+  auto queued = file_handle_->write(
+      consumable_fragment, header_block_.offsetToBody() + header_block_.bodySize(),
+      [this, sz, end_stream](absl::StatusOr<size_t> write_result) {
+        absl::MutexLock lock(&mu_);
+        cancel_action_in_flight_ = nullptr;
+        if (!write_result.ok() || write_result.value() != sz) {
```

Comment:
ping on this one.

Reply:
There is no retrying for incomplete writes at this time. That's not changed by this PR. It could merit a separate issue if you think it's important to handle that.

Reply:
(My opinion is it's fine for a cache write to fail if an interrupt happened during one of its write operations.)

Reply:
That makes sense once the other chain of things to get a common filesystem mock underpinning all this is done. :)